miopen_dialect_idea.md

```diff
@@ -156,20 +179,190 @@ An example based on NCHW/KCYX/NKHW:
+      names = ["gemmN"],
+      transformation = "merge",
+      source_dimensions = [0, 2, 3],
+      source_names = ["n", "ho", "wo"]
+    }
+  ],
+  output_layout = ["gemmM", "gemmN"],
+  source_layout = ["n", "ko", "ho", "wo"]
+} : memref<?x?x?x?xf32> to memref<?x?xf32>
+```
+
+
+```mlir
+// input tensor
+%input_n_c_hipad_wipad = miopen.transform(%input) {
+  layout = [
+    {
+      dimensions = [0],
+      names = ["n"],
+      transformation = "passthorugh",
+      source_dimensions = [0],
+      source_names = ["n"]
+    },
+    {
+      dimensions = [1],
+      names = ["c"],
+      transformation = "passthorugh",
+      source_dimensions = [1],
+      source_names = ["c"]
+    },
+    {
+      dimensions = [2],
+      names = ["hipad"],
+      transformation = "pad",
+      parameters = [0, 0],
+      source_dimensions = [2],
+      source_names = ["hi"]
+    },
+    {
+      dimensions = [3],
+      names = ["wipad"],
+      transformation = "pad",
+      parameters = [0, 0],
+      source_dimensions = [3],
+      source_names = ["wi"]
+    }
+  ],
+  output_layout = ["n", "c", "hipad", "wipad"],
+  source_layout = ["n", "c", "h", "w"]
+} : memref<?x?x?x?xf32> to memref<?x?x?x?xf32>
+
+%input_n_c_y_ho_x_wo = miopen.transform(%input_n_c_hipad_wipad) {
+  layout = [
+    layout = [
+      {
+        dimensions = [0],
+        names = ["n"],
+        transformation = "passthrough",
+        source_dimensions = [0],
+        source_names = ["n"]
+      },
+      {
+        dimensions = [1],
+        names = ["c"],
+        transformation = "passthrough",
+        source_dimensions = [1],
+        source_names = ["c"]
+      },
+      {
+        dimensions = [2, 3],
+        names = ["y", "ho"],
+        transformation = "embed",
+        parameters = [2, [1, 1, 0]],
+        source_dimensions = [2],
+        source_names = ["hipad"]
+      },
+      {
+        dimensions = [4, 5],
+        names = ["x", "wo"],
+        transformation = "embed",
+        parameters = [2, [1, 1, 0]],
+        source_dimensions = [2],
+        source_names = ["wipad"]
+      }
+  ],
+  intermediate_layout = ["n", "c", "hipad", "wipad"]
+  output_layout = ["n", "c", "y", "ho", "x", "wo"]
+} : memref<?x?x?x?xf32> to memref<?x?x?x?x?x?x?xf32>
+
+%input_gemmK_gemmN = miopen.transform(%input_n_c_y_ho_x_wo) {
+  layout = [
+    {
+      dimensions = [0],
+      names = ["gemmK"],
+      transformation = "merge",
+      source_dimensions = [1, 2, 4],
+      source_names = ["c", "y", "x"]
+    },
+    {
+      dimensions = [1],
+      names = ["gemmN"],
+      transformation = "merge",
+      source_dimensions = [0, 3, 5],
+      source_names = ["n", "ho", "wo"]
+    }
+  ],
+  intermediate_layout = ["n", "c", "y", "ho", "x", "wo"],
+  output_layout = ["gemmK", "gemmN"]
+} : memref<?x?x?x?x?x?x?xf32> to memref<?x?xf32>
+```
+
+```mlir
+// apply gridwise GEMM
+miopen.gridwise_gemm(%filter_gemmK_gemmM, %output_gemmM_gemmN, %input_gemmK_gemmN) {
```

Comment:
is it a valid matrix multiplication to do  transpose(K*M) * (M*N) ?

Reply:
Updated to `transpose(K, M) * (K, N) = (M, N)`